### PR TITLE
Support for the bulkImport operation

### PR DESCRIPTION
**Pull request by ftomassetti**
With this PR we introduce a new operation, intended to import into the repository a large number of new nodes. The operation can be performed on an existing repository, in parallel with normal operations.

The goal is to make this operation much faster, with respect to the standard store operation.

This is done:
- By reducing checks (but enforcing that all nodes are new)
- By using the Copy operation

In addition to this we support providing the payload through JSON, ProtoBuf, or Flatbuffers. While the formats have very different speeds on serializing (see https://github.com/LionWeb-io/lionweb-java/pull/160 for details), the overall performance of the system is not changing that much based on the format used (see _Experiment_ section below). However, when using JSON we can send payloads of 100,000 nodes, but when trying with 500,000 (compress or not), the operation fails.

The files generated for protobuffers and flatbuffers are committed.

# Experiment

I tried running an application that parses a bunch of files written in a legacy language and stores the resulting ASTs in the LionWeb Repository. I stopped when, after processing an AST, I passed the threshold of 500,000 nodes.

I first tried using the normal insert, which stores the individual AST after each file has been parsed. After around 200 files, we reached the threshold.

It took a total of 3,658,025 ms (~61 minutes).

This resulted in a database containing:
* 500409 nodes
* 548546 containments
* 843276 properties
* 1144738 references
* 0 reserved ids
* 407 versions

I then tried using the bulk operation. I parsed several files until the insertion threshold was reached, at which point I stored multiple ASTs at once. I tried with insertion thresholds of 100,000 nodes and 500,000 nodes. In the first case, we did 5 insertions per run, in the second one, just one.

Using bulkImport with JSON, with an insertion threshold of 100,000, a total nodes threshold of 500,000, and no compression, it took 738213 ms (~12 minutes). The database's dimension was the same, but the number of versions was lower (3 instead of 407).

Other combinations I tried were slightly slower, with one exception: using Flatbuffers, with an insertion threshold of 500,000, a total nodes threshold of 500,000, and no compression. That took 735961ms. **This is the best result so far, and it means reducing the current insertion time by 5**. 

Note also that using an insertion threshold of 500,000 nodes with JSON, either with or without compression, would cause the sender to crash.

# Use case I am working with

In this experiment, we stop after processing ~200 files out of a codebase of 2,700 files. Currently, inserting the nodes for the entire codebase would take around 13 hours. We have worked with a codebase up to 15 times larger than that (~30M lines of code). Processing such a codebase could take around 195 hours.

With this change, we can reduce time by five, to 2.5 hours for this codebase and 39 hours for the larger codebase we could work with.

This performance could be acceptable for the current codebase we are working with, while it would still be difficult to work with very large codebases. We hope that with further optimizations, we can gain more performance improvements.

# Which formats should we support? 

One could argue that it does not make sense to support 3 formats, considering that so far, we have not gotten a sufficient performance boost to justify the extra effort. However, I think it is useful to support both JSON plus a format that supports very large insertions. One between flatbuffers and protobuf could be dropped, but I would prefer to do that after we tried more optimizations and ran more experiments. At that point, we could decide which one can be the fastest 

# What next?

For this to be usable in practice, we would also need to release a new version of the client (I have the code for LionWeb-Kotlin already written, as I used it for this experiment).

I would then continue working on possible optimizations, to bring the execution time further down.

### CODE REVIEW

**Comment by joswarmer** [comment]
Pushed a commit to set all imports using .js postfix.  And updated the tsconfig file to model-NodeNext, so this is checked by the typescript compiler.

File: packages/additionalapi/src/database/ImportLogic.ts [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    nodes.forEach(node => {
+        read_stream_string.push(node.id);
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push(node.classifier.language);
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push(node.classifier.version);
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push(node.classifier.key);
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push("{}");
```
Comment (ftomassetti) [inline packages/additionalapi/src/database/ImportLogic.ts:26]: Here I should list the actual annotations...

File: packages/additionalapi/src/database/ImportLogic.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        const classifier = node.classifier();
+        read_stream_string.push(node.id());
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push(classifier.language());
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push(classifier.version());
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push(classifier.key());
+        read_stream_string.push(SEPARATOR);
+        read_stream_string.push("{}");
```
Comment (ftomassetti) [inline packages/additionalapi/src/database/ImportLogic.ts:124]: Here I should list the actual annotations...

File: packages/additionalapi/src/database/ImportLogic.ts (hunk 1 above)
Comment (ftomassetti) [reply packages/additionalapi/src/database/ImportLogic.ts]: Done

File: packages/additionalapi/src/database/ImportLogic.ts (hunk 2 above)
Comment (ftomassetti) [reply packages/additionalapi/src/database/ImportLogic.ts]: Done

**Comment by ftomassetti** [comment]
Is there anything that I could do to improve this PR?

**Review by joswarmer — APPROVED** [review]
The code is looking good, but I lack proper knowledge of protobuf and flatbuffers so I cannot judge the details.
At a later stage we should add some tests to ruin in the CI, but we probably need a (TypeScript) client for this.

**Comment by ftomassetti** [comment]
> At a later stage we should add some tests to ruin in the CI, but we probably need a (TypeScript) client for this.

I have tested it with a client based on Kotlin, but yes, we definitely need tests for this within this project. I created a task to track that (#85)

File: packages/additionalapi/README.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# How to generate FlatBuffers code
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply packages/additionalapi/README.md:1]: I added more information on the README. I start by explaining what the Bulk Import it is for, and then explaining why we use FlatBuffers and ProtoBuf

File: packages/additionalapi/src/flatbuffers/chunk.fbs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+root_type FBChunk;
+
+/* Specific for Bulk Import */
+
+table FBBulkImport {
+    attachPoints: [FBAttachPoint];
+    nodes:[FBNode];
+}
+
+table FBAttachPoint {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply packages/additionalapi/src/flatbuffers/chunk.fbs:59]: I will add a definition in the README

File: packages/additionalapi/src/database/ImportLogic.ts [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                read_stream_string.push(SEPARATOR);
+                read_stream_string.push(node.id());
+                read_stream_string.push("\n");
+        }
+    }
+        read_stream_string.push(null);
+    return read_stream_string;
+}
+
+async function pipeInputIntoQueryStream(client: PoolClient, query: string, inputStream: Duplex, opDesc: string) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply packages/additionalapi/src/database/ImportLogic.ts]: Done

**Comment by ftomassetti** [comment]
Thank you @joswarmer !

**Pull request merged**